dora/core/common/src/main/java/alluxio/membership/AlluxioEtcdClient.java

```diff
@@ -0,0 +1,557 @@
+import io.etcd.jetcd.watch.WatchResponse;
+import io.netty.util.internal.StringUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+
+/**
+ * Wrapper class around jetcd client to achieve utilities API to talk with ETCD.
+ */
+public class AlluxioEtcdClient implements Closeable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AlluxioEtcdClient.class);
+  private static final Lock INSTANCE_LOCK = new ReentrantLock();
+  public static final long DEFAULT_LEASE_TTL_IN_SEC = 2L;
+  public static final long DEFAULT_TIMEOUT_IN_SEC = 2L;
+  public static final int RETRY_TIMES = 3;
+  private static final int RETRY_SLEEP_IN_MS = 100;
+  private static final int MAX_RETRY_SLEEP_IN_MS = 500;
+  @GuardedBy("INSTANCE_LOCK")
+  @Nullable
+  private static volatile AlluxioEtcdClient sAlluxioEtcdClient;
+  public final ServiceDiscoveryRecipe mServiceDiscovery;
+  private final AtomicBoolean mConnected = new AtomicBoolean(false);
+  private final Closer mCloser = Closer.create();
+  // only watch for children change(add/remove) for given parent path
+  private final ConcurrentHashMap<String, Watch.Watcher> mRegisteredWatchers =
+      new ConcurrentHashMap<>();
+  private Client mClient;
+  public String[] mEndpoints;
+
+  /**
+   * CTOR for AlluxioEtcdClient.
+   * @param conf
+   */
+  public AlluxioEtcdClient(AlluxioConfiguration conf) {
+    String clusterName = conf.getString(PropertyKey.ALLUXIO_CLUSTER_NAME);
+    List<String> endpointsList = conf.getList(PropertyKey.ETCD_ENDPOINTS);
+    mEndpoints = endpointsList.toArray(new String[0]);
+    mServiceDiscovery = new ServiceDiscoveryRecipe(this, clusterName);
+  }
+
+  /**
+   * Get the singleton instance of AlluxioEtcdClient.
+   * @param conf
+   * @return AlluxioEtcdClient
+   */
+  public static AlluxioEtcdClient getInstance(AlluxioConfiguration conf) {
+    if (sAlluxioEtcdClient == null) {
+      try (LockResource lockResource = new LockResource(INSTANCE_LOCK)) {
+        if (sAlluxioEtcdClient == null) {
+          sAlluxioEtcdClient = new AlluxioEtcdClient(conf);
+        }
+      }
+    }
+    return sAlluxioEtcdClient;
+  }
+
+  /**
+   * Create jetcd grpc client no forcing.
+   */
+  public void connect() {
+    connect(false);
+  }
+
+  /**
+   * Create jetcd grpc client with choice of force or not.
```

Comment:
nit: add a comment on what `force` means for a caller with no context.

Just a nit cuz i think the code is not too hard to understand.